l1-contracts/contracts/common/libraries/DynamicIncrementalMerkleMemory.sol

```diff
@@ -113,8 +114,71 @@ library DynamicIncrementalMerkleMemory {
         }
         // Note this is overloading the sides array with the root.
         self._sides[levels] = currentLevelHash;
+        self._needsRootRecalculation = false;
         return (index, currentLevelHash);
     }
+
+    /**
+     * @dev Insert a new leaf in the memory tree lazily. Returns the position of the inserted leaf in the
+     * tree. This is the lazy version that updates only the needed side array entry and defers
+     * root computation until root() is called.
+     *
+     * Hashing the leaf before calling this function is recommended as a protection against
+     * second pre-image attacks.
+     */
+    function pushLazy(Bytes32PushTree memory self, bytes32 leaf) internal pure returns (uint256 index) {
```

Comment:
do you think it would make sense, to have one main _pushInner function, which both push and pushLazy can call? It would reduce code duplication massively